helpers/testsuite/test_certificates.go

```diff
@@ -0,0 +1,177 @@
+// Functions which allow for the creation of dummy certificates, chains, and
+// keys from certificate requests.
+
+package testsuite
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+
+	"github.com/cloudflare/cfssl/csr"
+)
+
+// CreateCertificateChain creates a chain of certificates from a slice of
+// requests. The first request is the root certificate and the last is the
+// leaf. The chain is returned as a slice of PEM-encoded bytes.
+func CreateCertificateChain(requests []csr.CertificateRequest) (certChain []byte, key []byte, err error) {
+	// Create the root certificate using the first request. This will be
+	// self-signed.
+	certChain = make([]byte, 0)
+	rootCert, prevKey, err := CreateSelfSignedCert(requests[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	certChain = append(certChain, rootCert...)
+
+	// For each of the next requests, create a certificate signed by the
+	// previous certificate.
+	prevCert := rootCert
+	for _, request := range requests[1:] {
+		cert, key, err := SignCertificate(request, prevCert, prevKey)
+		if err != nil {
+			return nil, nil, err
+		}
+		certChain = append(certChain, byte('\n'))
+		certChain = append(certChain, cert...)
+		prevCert = cert
+		prevKey = key
+	}
+
+	return certChain, key, nil
+}
+
+// CreateSelfSignedCert creates a self-signed certificate from a certificate
+// request. This function just calls the CLI "gencert" command.
+func CreateSelfSignedCert(request csr.CertificateRequest) (encodedCert, encodedKey []byte, err error) {
+	// Marshall the request into JSON format and write it to a temporary file.
+	jsonBytes, err := json.Marshal(request)
+	if err != nil {
+		return nil, nil, err
+	}
+	tempFile, err := createTempFile(jsonBytes)
+	if err != nil {
+		os.Remove(tempFile)
+		return nil, nil, err
+	}
+
+	// Create the certificate with the CLI tools.
+	command := exec.Command("cfssl", "gencert", "-initca", tempFile)
```

Comment:
this is sort of complicated. why not using initca.New(request) directly? That's much easier, right?